### PR TITLE
[NNC] Add utility functions to check channels-last contiguous

### PR DESCRIPTION
**Pull request by EikanWang**
The `Buf` uses `std::vector<ExprHandle>` to represent its strides. The `ExprHandle` could be an immediate value or a mathematical expression with variables involved both for the static shape and dynamic shape. So it is hard to directly deduce the channels-last contiguous layout based on the numerical calculation. Hence, the utility functions of this PR are based on the pattern match to check whether the `Buf` is channels-last contiguous.

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/expr.h [hunk 1]
```diff
@@ -295,6 +302,50 @@ class TORCH_API Buf : public ExprNode<Buf> {
     return true;
   }
 
+  bool is_contigous(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:305]: Typo: contig**u**ous

File: torch/csrc/jit/tensorexpr/expr.h [hunk 2]
```diff
@@ -295,6 +302,50 @@ class TORCH_API Buf : public ExprNode<Buf> {
...
+      at::MemoryFormat memory_format = at::MemoryFormat::Contiguous) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:306]: I think we can mark this function as `const` and also move its implementation to `expr.cpp`.

File: torch/csrc/jit/tensorexpr/expr.h [hunk 3]
```diff
@@ -295,6 +302,50 @@ class TORCH_API Buf : public ExprNode<Buf> {
...
+      auto pre_dim = dim_order[i - 1];
+      res &= is_cont_with(cur_dim, pre_dim);
+      if (!res)
+        return false;
+    }
+
+    return true;
+  }
+
+  bool is_channels_last_1d_contiguous() {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:338]: This also can be marked as `const`. Also, could you please add a brief comment of what this function does?

File: torch/csrc/jit/tensorexpr/expr.h [hunk 4]
```diff
@@ -313,6 +364,13 @@ class TORCH_API BufHandle : public ExprHandle {
       Dtype dtype)
       : ExprHandle(Buf::make(name_hint, dims, dtype)) {}
 
+  BufHandle(
+      const std::string& name_hint,
+      const std::vector<ExprHandle>& dims,
+      const std::vector<ExprHandle>& strides,
+      Dtype dtype)
+      : ExprHandle(Buf::make(name_hint, dims, strides, dtype)) {}
+
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:373]: Let's also expose `is_contiguous` and `is_channels_last_1d_contiguous` here in `BufHandle` to keep its API in sync with API of `Buf`. In the end, it's just a simple wrapper.

File: test/cpp/tensorexpr/test_expr.cpp [hunk 5]
```diff
@@ -41,6 +42,121 @@ TEST(Expr, BasicValueTest02) {
...
+
+  shapGenInfo cont_shape_conf = {
+      {3, {{0, 1, 2}}}, {4, {{0, 1, 2, 3}}}, {5, {{0, 1, 2, 3, 4}}}};
+
+  auto shape_gen_fn = [vars](
+                          int ndims, shapGenInfo shape_gen_info) -> shapeInfo {
+    std::vector<ExprHandle> dims_expr_vec;
+    for (int i = 0; i < ndims; i++) {
+      dims_expr_vec.push_back(vars[i]);
+    }
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_expr.cpp:89]: Nit: we probably can generate this vector once and then just reuse it everywhere.

**Review by ZolotukhinM — APPROVED** [review]
This looks good! I have only some minor remarks below:

File: test/cpp/tensorexpr/test_expr.cpp (hunk 5 above)
Comment (EikanWang) [reply test/cpp/tensorexpr/test_expr.cpp]: It depends on the `ndims` . I polished it a bit.

File: torch/csrc/jit/tensorexpr/expr.h (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/expr.h]: Oh, yes. Thanks :)

**Comment by EikanWang** [comment]
@navahgar , @huiguoo , @IvanKobzarev any comments on this PR?

**Comment by navahgar** [comment]
> @navahgar , @huiguoo , @IvanKobzarev any comments on this PR?

@EikanWang I didn't review this since I noticed that @ZolotukhinM already approved it. It should be good enough if one of us approve it. I believe Mikhail is already looking to land this internally.

**Pull request closed without merging**